cli/sdo/voucher.go

```diff
@@ -0,0 +1,310 @@
+package sdo
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/open-horizon/anax/cli/cliutils"
+	"github.com/open-horizon/anax/cli/exchange"
+	"github.com/open-horizon/anax/config"
+	"github.com/open-horizon/anax/externalpolicy"
+	"github.com/open-horizon/anax/i18n"
+	"github.com/open-horizon/anax/persistence"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+)
+
+// Sub-commands for inspecting and importing an Intel SDO ownership voucher for an SDO device.
+
+// Structs for parsing the voucher
+type OhStruct struct {
+	R []interface{} `json:"r"`
+	Guid []byte `json:"g"` // making it type []byte will automatically base64 decode the json value
+	DeviceType string `json:"d"`
+}
+type Voucher struct {
+	Oh OhStruct `json:"oh"`
+}
+
+// Structs for the inspect output
+type InspectDevice struct {
+	Uuid string `json:"uuid"`
+	DeviceType string `json:"deviceType"`
+}
+type InspectVoucher struct {
+	RendezvousUrls []string `json:"rendezvousUrls"`
+}
+type InspectOutput struct {
+	Device InspectDevice `json:"device"`
+	Voucher InspectVoucher `json:"voucher"`
+}
+
+// hzn voucher inspect <voucher-file>
+func VoucherInspect(voucherFile *os.File) {
+	defer voucherFile.Close()
+	cliutils.Verbose("Inspecting voucher file name: %s", voucherFile.Name())
+	msgPrinter := i18n.GetMessagePrinter()
+
+	outStruct := InspectOutput{}
+	voucherBytes, err := ioutil.ReadAll(bufio.NewReader(voucherFile))
+	if err != nil {
+		cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("reading the bytes from %s: %v", voucherFile.Name(), err))
+	}
+	if err = parseVoucherBytes(voucherBytes, &outStruct); err != nil {
+		cliutils.Fatal(cliutils.JSON_PARSING_ERROR, msgPrinter.Sprintf("parsing the json from %s: %v", voucherFile.Name(), err))
+	}
+
+	output := cliutils.MarshalIndent(outStruct, "voucher inspect")
+	fmt.Println(output)
+}
+
+func parseVoucherBytes(voucherBytes []byte, outStruct *InspectOutput) error {
+	// Read the voucher file and parse as json
+	voucher := Voucher{}
+	if err := json.Unmarshal(voucherBytes, &voucher); err != nil {
+		return errors.New("parsing json: " + err.Error())
+	}
+
+	// Do further parsing of the json, for those parts that have varying types
+	// The json in this section looks like: [ 1, [ 4, { "dn": "RVSDO", "po": 8040, "pow": 8040, "pr": "http" } ] ]
+	for _, thing := range voucher.Oh.R {
+		switch t := thing.(type) {
+		case []interface{}:
+			for _, thing2 := range t {
+				switch t2 := thing2.(type) {
+				case map[string]interface{}:
+					//fmt.Printf("thing2: %v\n", t2)
```

Comment:
Can you get rid of these debug statements?

Reply:
Yes, sorry